Refactor bloom filter targeting for xthins and make it optional

ptschip opened the pull request:
A few performance improvements for the bloom filter targeting process as well as making it an optional feature defaulting to OFF.  In the past when we were still on the Bitcoin Core chain this and the mempools were constantly full we needed this feature to keep the bloom filters from getting too big, however it does incur significant overhead in generating the filter, particularly during giga net testing.  But now that the mempools are no longer full, we can make it optional.

ptschip committed:
Reduce the scope of cs_main locking in bloom filter targeting

ptschip committed:
Move code for bloom targeting into the bloom targeting code block.

This is a move only, in preparation for being able to toggle
the bloom targeting process.

ptschip committed:
Use excessiveBlockSize as the largest block possible when targeting.

ptschip committed:
Add command line argument use-bloom-filter-targeting

Allows us to toggle this value and leave it in the off state by default.
This feature was usuful in the past when we were running on the Bitcoin
Core chain and the mempool was often completely full. Using bloom filter
targeting in that case was very helpful in reducing the size of an
xthin, however, now that the mempool is completely cleared on the Cash
chain, we can leave this feature turned off by default.

If bloom target is not on then simply add all the mempool hashes to the filter.

ptschip committed:
Add a node that has bloom filter targeting enabled to thinblocks.py

sickpig commented on src/thinblock.cpp:
```diff
@@ -1695,28 +1691,21 @@ void BuildSeededBloomFilter(CBloomFilter &filterMemPool,
             {
-                CTransaction tx = mi->GetTx();
-
-                if (!IsFinalTx(tx, nHeight, nLockTimeCutoff))
```
does this mean that we are going to have a degradation in Xthin performances as soon as the fraction of mempool non final txns will start to increase?

ptschip replied:
it just means the hashes will all get added to the bloom filter, but you know, the acutally performance issue wasn't here anyway, it was a couple of lines above where we didn't take a reference to the txn ...this was fixed in a later commit, so I'll just drop this commit and leave it as it was before.

sickpig commented on src/thinblock.cpp:
```diff
@@ -1777,6 +1777,13 @@ void BuildSeededBloomFilter(CBloomFilter &filterMemPool,
+        mempool.queryHashes(vMempoolHashes);
+        setHighScoreMemPoolHashes.insert(vMempoolHashes.begin(), vMempoolHashes.end());
+    }
+
```
won't change anything, but wouldn't it be logic if this commit got merged w/ the previous one? I mean, in case this PR will be merged and not squashed by @gandrewstone, leaving the structure of the commits as is would break `git bisect` in case we need to use it to hunt a bug.

ptschip replied:
sure i can do that

ptschip committed:
Use a reference rather than copying tx and hash when doing bloom filter targeting.

ptschip committed:
Re-enable bloom filter targeting but only if mempool is too large.

Still alow the bloom targeting to be toggled and leave it on by default
but don't do the targeting unless the mempool is actually larger than EB.
That way we can still make use of the targeting if/when the mempools
get overfull.

ptschip commented:
@sickpig added one commit, which was to still leave the targeting on by default but just don't use it unless the mempool is large.
